CUTENSOR: axpy! and axpby! not mutating fixed

yapanuwan opened the pull request:
`axpy!` and `axpby!` routines are supposed to overwrite the input array (mutating / in-place operation) without allocating a new array in memory. From the official [LinearAlgebra documentation](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.axpy!),

> `axpy!(a, X, Y)`
> Overwrite Y with X*a + Y, where a is a scalar. Return Y.

This quick fix corrects the behavior as well as improving performance.

Luckily, according to NVIDIA's cuTENSOR documentation for [cutensorElementwiseBinary()](https://docs.nvidia.com/cuda/cutensor/api/cutensor.html#cutensorelementwisebinary), the output tensor could be aliasing an input tensor.

> Notice that D may alias any input tensor if they share the same memory layout (i.e., same tensor descriptor). 

So the fix turned out be really easy.

Please be warned that I'm a newbie when it comes to open-source collaborations. An expert's opinion is highly appreciated. I cross-checked the outputs in my own application. It worked correctly, and the performance had improved dramatically!

yapanuwan committed:
In-place multiplication

axpy! and axpby! are supposed to overwrite the input (in-place multiplication) without allocating a new CuTensor in memory. This fixes the incorrect behavior as well as improving performance.

yapanuwan committed:
Tests for in-place behavior

Creates a copy of input tensor before calling axpy! and axpby! and checks if the returned tensor points to the overwritten input.

yapanuwan commented:
UPDATE: Unit test `cutensor/elementwise_binary` passes with modifications to check for correct behavior.

maleadt commented on test/cutensor/elementwise_binary.jl:
```diff
@@ -130,12 +130,16 @@ eltypes = ((Float16, Float16),
 
         α = rand(eltyD)
-        ctD = LinearAlgebra.axpy!(α, ctA, ctC)
+        ctD = CuTensor(copy(ctC.data), copy(ctC.inds))
```
I don't think we need to make an actual copy here, but just assert that the returned value is equal (`===`) to the input object. That also doesn't compare actual memory, but just that the objects are the same.

yapanuwan replied:
Sorry for the late reply. But the test fails if we don't make a copy first because `axpy!` mutates `ctC` which has to be reused in the next test for `axpby!`. However, I did some refactoring which should make things cleaner and clearer.

maleadt commented on test/cutensor/elementwise_binary.jl:
```diff
@@ -130,12 +130,16 @@ eltypes = ((Float16, Float16),
 
         γ = rand(eltyD)
-        ctD = LinearAlgebra.axpby!(α, ctA, γ, ctC)
+        ctD = CuTensor(copy(ctC.data), copy(ctC.inds))
```
Same here.

yapanuwan replied:
We can get away without making a copy here, because this is the last test. But I still think it should be done to avoid future headaches and for consistency. What do you think?

maleadt reviewed (changes requested):
Thanks for the contribution, and sorry for the delayed response! LGTM, but I'd change the test to prevent needless copying and comparison.

yapanuwan committed:
Refractor by implementing CuTensor copying

Implement Base.copy() for CuTensors which duplicates the object with all its contents.
